src/executor/evaluate/mod.rs

```diff
@@ -601,6 +601,52 @@ async fn evaluate_function<'a, T: 'static + Debug>(
             }
         }
         .map(Evaluated::from),
+
+        Function::Substr { expr, start, count } => {
+            let string = match eval_to_str(expr).await? {
+                Nullable::Value(v) => v,
+                Nullable::Null => {
+                    return Ok(Evaluated::from(Value::Null));
+                }
+            };
+            let start = match eval_to_integer(start).await? {
+                Nullable::Value(v) => v,
+                Nullable::Null => return Ok(Evaluated::from(Value::Null)),
+            };
+
+            let count = match count {
+                Some(expr) => match eval_to_integer(expr).await? {
+                    Nullable::Value(v) => match v {
+                        x if x < 0 => return Err(EvaluateError::NegativeSubstrLenNotAllowed.into()),
+                        _ => Some(v),
+                    },
+                    Nullable::Null => None,
+                },
+                None => None,
+            };
+
+            let s: usize = if start <= 0 { 0 } else { (start - 1) as usize };
+            let e = match count {
+                Some(v) => {
+                    if (start - 1 + v) < 0 {
+                        0
+                    } else if (start - 1 + v) <= string.len() as i64 {
+                        (start - 1 + v) as usize
+                    } else {
+                        string.len()
+                    }
+                }
+                None => string.len(),
+            };
+
+            let result = if s >= string.len() {
+                String::from("")
+            } else {
+                String::from(&string[s..e])
+            };
```

Comment:
Ok, here type casting to `usize` looks inevitable.
However, we can simplify the codes by doing a bit of rearrangement.
`std::cmp::{min, max}` would be useful.

```rust                                                                    
let start = match eval_to_integer(start).await? {                             
    Nullable::Value(v) => v - 1,                                              
    Nullable::Null => {                                                       
        return Ok(Evaluated::from(Value::Null));                              
    }                                                                         
};                                                                            
                                                                              
let end = match count {                                                       
    Some(expr) => match eval_to_integer(expr).await? {                        
        Nullable::Value(count) => {                                           
            if count < 0 {                                                    
                return Err(EvaluateError::NegativeSubstrLenNotAllowed.into());
            }                                                                 
                                                                              
            min(max(start + count, 0) as usize, string.len())                 
        }                                                                     
        Nullable::Null => {                                                   
            return Ok(Evaluated::from(Value::Null));                          
        }                                                                     
    },                                                                        
    None => string.len(),                                                     
};                                                                            
                                                                              
let start = min(max(start, 0) as usize, string.len());                        
let string = String::from(&string[start..end]);                               
```

Reply:
I mainly focused on two points.
1. handling `start` codes were splitted into two.
```rust
let s: usize = if start <= 0 { 0 } else { (start - 1) as usize };
```
```rust
let result = if s >= string.len() {
    String::from("")
} else {
    String::from(&string[s..e])
};
```

2. Also the `count` handling match codes
```rust
let count = match count {
    Some(expr) => match eval_to_integer(expr).await? {
         ..
```
```rust
let e = match count {
    Some(v) => {
        if (start - 1 + v) < 0 {
    ..
```